driver-core/src/main/com/mongodb/internal/async/AsyncSupplier.java

```diff
@@ -0,0 +1,138 @@
+ */
+@FunctionalInterface
+public interface AsyncSupplier<T> extends AsyncFunction<Void, T> {
+    /**
+     * This should not be called externally to this API. It should be
+     * implemented as a lambda. To "finish" an async chain, use one of
+     * the "finish" methods.
+     *
+     * @see #finish(SingleResultCallback)
+     */
+    void unsafeFinish(SingleResultCallback<T> callback);
+
+    /**
+     * This method must only be used when this AsyncSupplier corresponds
+     * to a {@link java.util.function.Supplier} (and is therefore being
+     * used within an async chain method lambda).
+     * @param callback the callback
+     */
+    default void getAsync(final SingleResultCallback<T> callback) {
+        unsafeFinish(callback);
+    }
+
+    @Override
+    default void unsafeFinish(@Nullable final Void value, final SingleResultCallback<T> callback) {
+        unsafeFinish(callback);
+    }
+
+    /**
+     * Must be invoked at end of async chain.
+     * @param callback the callback provided by the method the chain is used in
+     */
+    default void finish(final SingleResultCallback<T> callback) {
+        final boolean[] callbackInvoked = {false};
+        try {
+            this.unsafeFinish((v, e) -> {
+                callbackInvoked[0] = true;
+                callback.onResult(v, e);
+            });
+        } catch (Throwable t) {
+            if (callbackInvoked[0]) {
+                throw t;
+            } else {
+                callback.onResult(null, t);
+            }
+        }
+    }
+
+    /**
+     * @param function The async function to run after this runnable
+     * @return the composition of this supplier and the function, a supplier
+     * @param <R> The return type of the resulting supplier
+     */
+    default <R> AsyncSupplier<R> thenApply(final AsyncFunction<T, R> function) {
+        return (c) -> {
+            this.unsafeFinish((v, e) -> {
+                if (e == null) {
+                    function.unsafeFinish(v, c);
+                } else {
+                    c.onResult(null, e);
+                }
+            });
+        };
+    }
+
+
+    /**
+     * @param consumer The async consumer to run after this supplier
+     * @return the composition of this supplier and the consumer, a runnable
+     */
+    default AsyncRunnable thenConsume(final AsyncConsumer<T> consumer) {
+        return (c) -> {
+            this.unsafeFinish((v, e) -> {
+                if (e == null) {
+                    consumer.unsafeFinish(v, c);
+                } else {
+                    c.onResult(null, e);
+                }
+            });
+        };
+    }
+
+    /**
+     * @param errorCheck A check, comparable to a catch-if/otherwise-rethrow
+     * @param supplier   The branch to execute if the error matches
+     * @return The composition of this, and the conditional branch
+     */
+    default AsyncSupplier<T> onErrorIf(
```

Comment:
Because we are removing code duplication using inheritance, async runnables are void suppliers (or, void-void functions). However, our naming uses run/supply, so for consistency, this method should be called `onErrorRunOrSupplyIf`, but this is confusing.

I think that the verb here is something like "handle" X, and "onX" is a conventional way to name such methods, which is at least less confusing than the above.

Reply:
> async runnables are void suppliers

This is irrelevant, as the `onErrorIf` method takes `AsyncSupplier`, i.e., it's `onErrorSupplyIf`.

P.S. This thread stays optional.

Reply:
Leaving as-is. `onErrorSupplyIf` is a confusing name when the thing being supplied is "nothing" (void).